Ensure value and defaultValue do not assign functions and symbols

nhunzaker opened the pull request:
I think this addresses https://github.com/facebook/react/issues/11734. I need to verify with the attribute table.

<img width="1269" alt="screen shot 2017-12-01 at 8 47 14 am" src="https://user-images.githubusercontent.com/590904/33485414-43621f8e-d674-11e7-813b-cc3410863a92.png">

nhunzaker commented on packages/react-dom/src/client/ReactDOMFiberInput.js:
```diff
@@ -314,13 +310,39 @@ function synchronizeDefaultValue(
+      // Use hasOwnProperty instead of checking for '' to ensure properties
+      // are set for the first time. This is important for defaultValue, which
+      // otherwise does not set an empty value attribute
+      if (!node.hasOwnProperty(property) || node[property] !== '' + value) {
```
We must only compare the stringified `value` property/attribute after we've confirmed that it isn't a symbol. Otherwise you get a run time exception:

<img width="370" alt="screen shot 2017-12-01 at 8 51 40 am" src="https://user-images.githubusercontent.com/590904/33485591-e87b1304-d674-11e7-8f0a-4f3feb26de53.png">

Maybe a safer way to do this would be to wrap the value in a string constructor?

<img width="227" alt="screen shot 2017-12-01 at 8 52 36 am" src="https://user-images.githubusercontent.com/590904/33485608-fddbc77a-d674-11e7-83f2-6f8af1bd8774.png">

<img width="362" alt="screen shot 2017-12-01 at 8 53 02 am" src="https://user-images.githubusercontent.com/590904/33485626-114ff678-d675-11e7-95c0-397194445712.png">

gaearon replied:
AFAIK `String()` is slower, at least in V8.

nhunzaker replied:
Sure enough:

![Uploading Screen Shot 2017-12-01 at 9.07.22 AM.png…]()

jquense replied:
Should we add a TODO: to remove this on a major bump? It's not clear to me why ignoring other types vs throwing or defering to whatever the browser does is more corrrect

gaearon replied:
I think the constraint we care about is that objects get stringified. Since we depend on this a lot. Therefore we can't use browser APIs that don't attempt to stringify them.

jquense replied:
Sure that's sense, tho I think "treating the as null" vs an error is maybe less ideal. In, the error case it's easier track down and fix vs inputs clearing and maybe not noticing

nhunzaker replied:
I believe the unknown property hook will still raise a warning.

nhunzaker commented:
The attribute table seems correct, though now I wonder if we should share this behavior with textarea and select:

<img width="1790" alt="screen shot 2017-12-01 at 8 57 18 am" src="https://user-images.githubusercontent.com/590904/33485798-c524104e-d675-11e7-8ad9-35d0cc834cf2.png">

**Attribute Table:**

[AttributeTableSnapshot.md.zip](https://github.com/facebook/react/files/1521803/AttributeTableSnapshot.md.zip)

gaearon commented on packages/react-dom/src/client/ReactDOMFiberInput.js:
```diff
@@ -314,13 +310,39 @@ function synchronizeDefaultValue(
+  value: *,
+) {
+  // TODO: This should use DOMProperty.shouldSetAttribute, however the current
+  // behavior is such that invalid attributes do not update the current attribute,
```
This sounds like a bug to me. I'd like to always treat them as `null` instead of showing stale values. I'd be fine changing behavior even if that's what we did in 16 because I think it's more likely to cause product bugs.

gaearon commented on packages/react-dom/src/client/ReactDOMFiberInput.js:
```diff
@@ -314,13 +310,39 @@ function synchronizeDefaultValue(
+        // Cast `value` to a string to ensure the value is set correctly. While
+        // browsers typically do this as necessary, jsdom doesn't.
+        // TODO: Is this still true with reasonably modern JSDOM?
+        node[property] = '' + value;
```
I'd expect dynamic property access to be slower than just setting a property normally.

nhunzaker commented:
> I don't mind changing behavior to "always reset to null for bad inputs" since keeping stale values seems worse to me.

I think that makes things a bit easier.

nhunzaker commented on packages/react-dom/src/events/ChangeEventPlugin.js:
```diff
@@ -228,17 +229,15 @@ function handleControlledInputBlur(inst, node) {
 
   // Fiber and ReactDOM keep wrapper state in separate places
+  // TODO: Is this still necessary now that Stack is gone?
   let state = inst._wrapperState || node._wrapperState;
```
Left a note to follow up here.

jquense replied:
this should not be necessary anymore

nhunzaker commented on packages/react-dom/src/shared/DOMProperty.js:
```diff
@@ -12,8 +12,11 @@ import warning from 'fbjs/lib/warning';
   defaultChecked: true,
+  // TODO: This prevents the assignment of defaultValue to regular
+  // elements (not just inputs). Now that ReactDOMInput assigns to the
+  // defaultValue property -- do we need this?
```
Left a note to follow up here.

nhunzaker commented on packages/react-dom/src/events/ChangeEventPlugin.js:
```diff
@@ -228,17 +229,15 @@ function handleControlledInputBlur(inst, node) {
-  if (node.getAttribute('value') !== value) {
-    node.setAttribute('value', value);
-  }
+  synchronizeDefaultValue(node, node.type, node.value);
```
@gaearon new here: I forgot about the number behavior on blur when we change inputs. This is so that chrome doesn't validate number inputs when we assign the default value. The DOM Operations are now consistent.

nhunzaker commented:
Pushed a bit of a work in progress pass. There's too much branching. I think we can eliminate a few conditions.

gaearon commented:
Let me know when you’re satisfied :-)

nhunzaker commented:
@gaearon pushed up some final bits. I'd appreciate a very critical eye. I'm a bit tapped for time to invest in this, but I can pick it up later today and answer any questions.

nhunzaker commented:
Basic browser testing checks out. It looks like this really does fix #8395:

<img width="1106" alt="screen shot 2017-12-01 at 10 39 30 am" src="https://user-images.githubusercontent.com/590904/33490104-fc074dde-d683-11e7-9cdc-a9328c82cdda.png">

gaearon commented on packages/react-dom/src/__tests__/ReactDOMInput-test.js:
```diff
@@ -1491,4 +1491,84 @@ describe('ReactDOMInput', () => {
+    });
+
+    it('does not update the value to a Symbol', function() {
+      var container = document.createElement('div');
```
Shouldn't it display a warning?

gaearon commented:
Mind rebasing now?

gaearon commented:
Or I could.

nhunzaker commented:
@gaearon I can get it if you haven't already started.

nhunzaker commented on packages/react-dom/src/__tests__/ReactDOMInput-test.js:
```diff
@@ -1585,4 +1585,116 @@ describe('ReactDOMInput', () => {
+      expect(node.value).toBe('');
+      expect(node.getAttribute('value')).toBe('');
+    });
+  });
```
@gaearon assigning defaultValue does not raise a warning. This is because `defaultValue` is considered a reserved prop:
https://github.com/facebook/react/blob/master/packages/react-dom/src/shared/DOMProperty.js#L12-L21

Removing it from the reserved props list addresses the issue, but then a _ton_ of server integration tests fail. The current 16.2.0 behavior also does not raise a warning:

<img width="1401" alt="screen shot 2017-12-01 at 12 50 20 pm" src="https://user-images.githubusercontent.com/590904/33495913-3b8bfc90-d696-11e7-9fdf-1eb9877d769f.png">

Do we want to address the lack of a warning for `defaultValue` in this PR?

gaearon replied:
No, it's okay.

aweary commented on packages/react-dom/src/client/ReactDOMFiberInput.js:
```diff
@@ -117,11 +118,17 @@ export function initWrapperState(element: Element, props: Object) {
+  var initialValue = props.value != null ? props.value : defaultValue;
   var node = ((element: any): InputWithWrapperState);
+
+  if (!shouldSetAttribute('value', initialValue)) {
```
Not a blocker, but it might be worth using a method specific to `value` since `shouldSetAttribute` does a [few unnecessary checks in this case](https://github.com/facebook/react/blob/master/packages/react-dom/src/shared/DOMProperty.js#L144-L153). It might also call `shouldAttributeAcceptBooleanValue` [which also does more work than needed](https://github.com/facebook/react/blob/master/packages/react-dom/src/shared/DOMProperty.js#L176-L188)

I don't think this is a really hot path, so it's probably fine the way it is, but just worth noting.

nhunzaker replied:
Yeah. I had an `assignProperty` method earlier, but we thought it was too much indirection. I'd be happy to revisit this though.

gaearon replied:
Let's fork `shouldSetAttribute` here to `shouldSetValue`? I'm not against a special method per se, I'm against a special method for doing the assignment.

nhunzaker commented:
Rebased with the ChangeEventPlugin on master. I've also added a specific method for getting a safe assignment value for `value` and `defaultValue`, as recommended by @aweary.

Flow still fails because it doesn't like a loose equality check between numbers and strings:

<img width="558" alt="screen shot 2017-12-02 at 12 23 57 pm" src="https://user-images.githubusercontent.com/590904/33517996-b0b5abd6-d75b-11e7-959f-0d68270d6f03.png">

Playing around with that code, now that we have a pretty comprehensive set of tests around numeric equality, I think I have a more concise solution that makes flow happy. I'm going to send it out as a PR against this branch.

nhunzaker commented:
https://github.com/facebook/react/pull/11751 addresses the Flow issue, which I think is the last outstanding thing here.
